src/features/courses/course-controller.ts

```diff
@@ -2522,6 +2634,243 @@ class CourseController {
             throw new WrappedError('Could not create new student grade', e);
         }
     }
+
+    // does this actually belong in course-repository?
+    async createStudentTopicAssessmentInfo(options: CreateNewStudentTopicAssessmentInfoOptions): Promise<StudentTopicAssessmentInfo> {
+        const { 
+            courseTopicContentId,
+            startTime,
+            endTime,
+            nextVersionAvailableTime,
+            maxAttempts
+        } = options;
+        try {
+            return await StudentTopicAssessmentInfo.create({
+                courseTopicContentId,
+                startTime,
+                endTime,
+                nextVersionAvailableTime,
+                maxAttempts
+            });
+        } catch (e) {
+            throw new WrappedError('Could not create new Student Topic Assessment info for this topic', e);
+        }
+    }
+
+    async createGradeInstancesForAssessment(options: CreateGradeInstancesForAssessmentOptions): Promise<StudentTopicAssessmentInfo> {
+        const { topicId, userId } = options;
+        const topic = await courseRepository.getCourseTopic({id: topicId});
+        const topicInfo = await topic.getTopicAssessmentInfo(); // order by startDate 
+
+        const questions = await courseRepository.getQuestionsFromTopicId({id:topicId});
+        const startTime = moment().add(1,'minute'); // 1 minute should cover any delay in creating records before a student can actually begin
+
+        let endTime = moment(startTime).add(topicInfo.duration, 'minutes');
+        if (topicInfo.hardCutoff) {
+            endTime = moment.min(moment(topic.deadDate), moment(startTime).add(topicInfo.duration, 'minutes'));
```

Comment:
We have a convenient extension toMoment
This could be:
endTime = moment.min(topic.deadDate.toMoment(), startTime.toMoment().add(topicInfo.duration, 'minutes'));

I think this looks cleaner
The main reason for the extension is for null chaining too